### ISSUE TITLE
Use Skylark for git_repository rules

### ISSUE THREAD

**Issue by davidzchen** [bazelbuild/bazel#1408]
There are many benefits for replacing the native `git_repository` rules that use jgit with Skylark rules that invoke the native Git binaries:

* Better support for cloning over SSH and from behind proxies
* Support for shallow git clones (which jgit does not support)

@nelhage has already written a set of Git repository rules in Skylark: https://github.com/nelhage/bazel_git_repositories. Perhaps it would be good to get them merged in to Bazel and replace the native rules.

**Comment by davidzchen** [comment]
Continuing the discussion from #810, @johnynek will be reaching out to @nelhage to discuss moving the Skylark `git_repository` rules into Bazel.

The Skylark rules work well as drop-in replacements for the native rules, but we will need #1289 in order to get the [integration test](https://github.com/bazelbuild/bazel/blob/master/src/test/shell/bazel/git_repository_test.sh) to pass since the tests check the progress messages of the rules.

**Comment by nelhage** [comment]
Hey folks,

I'd be happy to help get those rules upstream, if there's interest. We're
relying on them quite heavily already internally, and would love to have
them upstream.

I'm on vacation until this Thursday, but happy to spend time on this after
I'm back.

**Comment by davidzchen** [comment]
Hi @nelhage, we are definitely still interested in replacing the Java `git_repository` rules with your `bazel_git_repository` rules. Can you start a Gerritt review for getting the rules merged into Bazel? I think `tools/build_defs/repo` would be a good place to import the rules.

To get the Git integration tests to pass with the new rules, we would need #1289, but we can import the rules first and then make the change to the Git integration tests once #1289 is done. @damienmg - are you working on #1289? If not, I'd be happy to pick it up.

**Comment by nelhage** [comment]
@davidzchen I've opened https://bazel-review.googlesource.com/#/c/3982/ with an import of the current version from `bazel_git_repository`. I haven't imported any tests yet.

**Comment by davidzchen** [comment]
Awesome, thanks!

**Comment by danmane** [comment]
When this is merged, will there be any action required by TensorFlow maintainers to have TensorFlow use the native binaries? Or will it just be a seamless replacement?

**Comment by kchodorow** [comment]
Code will continue to use the native rules automatically.  If you'd like to try the Skylark ones, you can add:

```
load('@bazel_tools//tools/build_defs/repo:git_repositories.bzl', 'git_repository')
```

to your WORKSPACE file.

**Comment by kchodorow** [comment]
This hasn't been released yet, but was added by 247bbfa751d786d97839749a07d0c212478ffad7.

**Comment by david-german-tri** [comment]
For anyone else who Googles their way to this thread: since https://github.com/bazelbuild/bazel/issues/1408#issuecomment-240176593, the Skylark rules appear to have moved.  The `load` statement that worked for me is:

```
load('@bazel_tools//tools/build_defs/repo:git.bzl', 'git_repository')
```

[git_repositories.bzl](https://github.com/bazelbuild/bazel/commits/master/tools/build_defs/repo/git_repositories.bzl) attempts to redirect there, but has syntax errors: `Implicit string concatenation is forbidden`

**Comment by kamalmarhubi** [comment]
Should this have been closed? I don't think the current skylark git repository rules handle shallow cloning yet?

**Comment by davidzchen** [comment]
@kamalmarhubi Let's open a separate issue for shallow clone support?
